datalad/support/gitrepo.py

```diff
@@ -416,6 +418,213 @@ def wrapped(repo, *args, **kwargs):
     return wrapped
 
 
+class GitProgress(object):
+    """Reduced variant of GitPython's RemoteProgress class
+
+    Original copyright:
+        Copyright (C) 2008, 2009 Michael Trier and contributors
+    Original license:
+        BSD 3-Clause "New" or "Revised" License
+    """
+    _num_op_codes = 10
+    BEGIN, END, COUNTING, COMPRESSING, WRITING, RECEIVING, RESOLVING, FINDING_SOURCES, CHECKING_OUT, ENUMERATING = \
+        [1 << x for x in range(_num_op_codes)]
+    STAGE_MASK = BEGIN | END
+    OP_MASK = ~STAGE_MASK
+
+    DONE_TOKEN = 'done.'
+    TOKEN_SEPARATOR = ', '
+
+    _known_ops = {
+        COUNTING: ("Counting", "Objects"),
+        ENUMERATING: ("Enumerating", "Objects"),
+        COMPRESSING: ("Compressing", "Objects"),
+        WRITING: ("Writing", "Objects"),
+        RECEIVING: ("Receiving", "Objects"),
+        RESOLVING: ("Resolving", "Deltas"),
+        FINDING_SOURCES: ("Finding", "Sources"),
+        CHECKING_OUT: ("Check out", "Things"),
+    }
+
+    __slots__ = ('_seen_ops', '_pbars')
+
+    re_op_absolute = re.compile(r"(remote: )?([\w\s]+):\s+()(\d+)()(.*)")
+    re_op_relative = re.compile(r"(remote: )?([\w\s]+):\s+(\d+)% \((\d+)/(\d+)\)(.*)")
+
+    def __init__(self):
+        self.__enter__()
+
+    def __enter__(self):
+        self._seen_ops = []
+        self._pbars = set()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        # take down any progress bars that were not closed orderly
+        for pbar_id in self._pbars:
+            log_progress(
+                lgr.info,
+                pbar_id,
+                'Finished',
+            )
+
+    def __call__(self, byts):
+        """Callable interface compatible with WitlessRunner()
+
+        Parameters
+        ----------
+        byts : bytes
+          One or more lines of command output.
+
+        Returns
+        -------
+        bytes
+          All input in its original form, except for lines that
+          were identified as recognized progress reports.
+        """
+        keep_lines = []
+        for line in byts.splitlines(keepends=True):
+            if not self._parse_progress_line(line):
+                keep_lines.append(line)
+        # the zero indicated that no data remained unprocessed at the
+        # end of the input
+        # TODO reevaluate whether it is useful to keep this feature
```

Comment:
Considering you said this PR was ready to go in your view, I guess you've decided it was useful, and this comment should be removed and the docstring(s) updated (which I think Yarik mentioned).

I didn't spot a helper that returned something other than 0.  Why is the feature useful?

Reply:
It is useful whenever some output processor can only meaningfully process a sufficiently large chunk of data. For example, something expects JSON lines output, but for some reason only gets passed 1.5 lines (a full one and another line with an incomplete record -- who knows why, maybe a buffer setting in an external tool). With this feature an output processor can reject some input, and get it passed again later on. This avoids having to implement state-tracking in processors and might keep them simpler.

Reply:
Thanks for the explanation.   Makes sense and it's fine by me, though it does feel a bit premature to add it when no current callers use it.